File: examples3d/convex_polyhedron3.rs
use rand::distr::{Distribution, StandardUniform};
use rand::{SeedableRng, rngs::StdRng};
use rapier_testbed3d::Testbed;
use rapier3d::prelude::*;

pub fn init_world(testbed: &mut Testbed) {
    /*
     * World
     */
    let mut bodies = RigidBodySet::new();
    let mut colliders = ColliderSet::new();
    let impulse_joints = ImpulseJointSet::new();
    let multibody_joints = MultibodyJointSet::new();

    /*
     * Ground
     */
    let ground_size = 40.0;
    let ground_height = 0.1;

    let rigid_body = RigidBodyBuilder::fixed().translation(vector![0.0, -ground_height, 0.0]);
    let handle = bodies.insert(rigid_body);
    let collider = ColliderBuilder::cuboid(ground_size, ground_height, ground_size);
    colliders.insert_with_parent(collider, handle, &mut bodies);

    /*
     * Create the polyhedra
     */
    let num = 5;
    let scale = 2.0;
    let border_rad = 0.1;

    let shift = border_rad * 2.0 + scale;
    let centerx = shift * (num / 2) as f32;
    let centery = shift / 2.0;
    let centerz = shift * (num / 2) as f32;

    let mut rng = StdRng::seed_from_u64(0);
    let distribution = StandardUniform;

    for j in 0usize..25 {
        for i in 0..num {
            for k in 0usize..num {
                let x = i as f32 * shift - centerx;
                let y = j as f32 * shift + centery + 3.0;
                let z = k as f32 * shift - centerz;

                let mut points = Vec::new();
                for _ in 0..10 {
                    let pt: Point<f32> = distribution.sample(&mut rng);
                    points.push(pt * scale);
                }

                // Build the rigid body.
                let rigid_body = RigidBodyBuilder::dynamic().translation(vector![x, y, z]);
                let handle = bodies.insert(rigid_body);
                let collider = ColliderBuilder::round_convex_hull(&points, border_rad).unwrap();
                colliders.insert_with_parent(collider, handle, &mut bodies);
            }
        }
    }

    /*
     * Set up the testbed.
     */
    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints);
    testbed.look_at(point![30.0, 30.0, 30.0], Point::origin());
}

File: python/rapier-testbed/rapier_testbed/examples3/convex_polyhedron3.py
"""Port of ``examples3d/convex_polyhedron3.rs``.

Random clouds of 10 points each are turned into convex polyhedra via
``ColliderBuilder::convex_hull`` (the Rust example uses
``round_convex_hull`` which is not exposed in the Python bindings yet,
so we fall back to the plain convex hull).
"""
from __future__ import annotations

import random

import numpy as np

import rapier3d as rp

from .._registry import register

CATEGORY = "Collisions"
NAME = "Convex polyhedron"


def init_world(testbed) -> None:
    bodies = rp.RigidBodySet()
    colliders = rp.ColliderSet()
    impulse_joints = rp.ImpulseJointSet()
    multibody_joints = rp.MultibodyJointSet()

    ground_size = 40.0
    ground_height = 0.1
    ground_body = rp.RigidBody.fixed().translation((0.0, -ground_height, 0.0))
    ground_handle = bodies.insert(ground_body)
    colliders.insert_with_parent(
        rp.Collider.cuboid(ground_size, ground_height, ground_size),
        ground_handle,
        bodies,
    )

    num = 5
    scale = 2.0
    border_rad = 0.1
    shift = border_rad * 2.0 + scale
    centerx = shift * (num // 2)
    centery = shift / 2.0
    centerz = shift * (num // 2)

    rng = random.Random(0)
    for j in range(25):
        for i in range(num):
            for k in range(num):
                x = i * shift - centerx
                y = j * shift + centery + 3.0
                z = k * shift - centerz

                pts = np.array(
                    [
                        [rng.random() * scale, rng.random() * scale, rng.random() * scale]
                        for _ in range(10)
                    ],
                    dtype=np.float32,
                )

                body = rp.RigidBody.dynamic().translation((x, y, z))
                handle = bodies.insert(body)
                try:
                    coll = rp.Collider.convex_hull(pts)
                except Exception:
                    # Degenerate point cloud; skip.
                    continue
                colliders.insert_with_parent(coll, handle, bodies)

    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints)
    testbed.look_at((30.0, 30.0, 30.0), (0.0, 0.0, 0.0))


register(CATEGORY, NAME, init_world, dim=3)


if __name__ == "__main__":
    from .._picker import run

    run(initial=f"{CATEGORY} / {NAME}")
